xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-templates/src/main/resources/templates/notification/macros.vm

```diff
@@ -243,3 +234,45 @@ $xwiki.getPlainUserName($user)##
   #end
   #displayNotificationEventSkeleton($icon, 'comment', $content, '')
 #end
+
+#macro(displayCompositeEventDescription $compositeEvent)
+  #set ($baseTranslationPrefix = "${compositeEvent.type}")
+  #set ($fallbackTranslationPrefix = "notifications.events.${compositeEvent.type}")
+  #set ($fallbackTranslationSuffix = "description")
+  #set ($fallbackTranslation = "#getNotificationTranslation($baseTranslationPrefix,$fallbackTranslationSuffix,$fallbackTranslationPrefix,[])")
+  #set ($descriptionTranslation = $NULL)
+  #if ($compositeEvent.users.size() == 1)
+    #set ($translationSuffix = "description.by.1user")
+    #set ($translationParameters = ["#displayNotificationEventUsers($compositeEvent.users, true, false)"])
+    #set ($descriptionTranslation = "#getNotificationTranslation($baseTranslationPrefix,$translationSuffix,$fallbackTranslationPrefix,$translationParameters)")
+    #if ("$!descriptionTranslation" != '')
+      $descriptionTranslation
+    #else
+      $fallbackTranslation
+    #end
+  #else
+  <div>
+    #set ($translationSuffix = "description.by.users")
+    #set ($translationParameters = [$compositeEvent.users.size()])
+    #set ($descriptionTranslation = "#getNotificationTranslation($baseTranslationPrefix,$translationSuffix,$fallbackTranslationPrefix,$translationParameters)")
+    #if ("$!descriptionTranslation" != '')
+      $descriptionTranslation
+    #else
+      $fallbackTranslation
+    #end
+  </div>
+  <div>
+    #displayNotificationEventUsers($compositeEvent.users, true, true)
+  </div>
+  #end
+#end
+
+#macro(getNotificationTranslation $prefix $suffix $fallbackPrefix $parameters)
+  #set ($translationResult = $NULL)
+  #set ($translationKey = "${prefix}.${suffix}")
+  #set ($translationResult = $services.localization.render($translationKey, $parameters))
```

Comment:
You can pass directly several keys to #render instead of taking care of the fallback yourself.

Reply:
Wow, I wasn't aware of that API... definitely easier to implement my solution.

Reply:
If we pass a parameter to a translation that doesn't use it, it's just ignoring the parameter, right?

Reply:
> If we pass a parameter to a translation that doesn't use it, it's just ignoring the parameter, right?

Yes, the only constraint when you use this API is that all the keys must expect compatible parameters but too many parameters is not a problem.